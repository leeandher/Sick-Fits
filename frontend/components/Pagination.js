import React from 'react'
import PaginationStyles from './styles/PaginationStyles'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'

import { PER_PAGE } from '../config'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`

const Pagination = ({ page }) => (
  <PaginationStyles>
    <Query query={PAGINATION_QUERY}>
      {({ data, error, loading }) => {
        if (loading) return <p>⚡ Loading... ⚡</p>
        if (error) return <p>❌ Error ❌: {error.message}</p>
        const { count } = data.itemsConnection.aggregate
        const pages = Math.ceil(count / PER_PAGE)
        return (
          <p>
            Page {page} of {pages}
          </p>
        )
      }}
    </Query>
  </PaginationStyles>
)

export default Pagination
