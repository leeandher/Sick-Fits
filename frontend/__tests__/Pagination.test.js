import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import Router from 'next/router'

import Pagination, { PAGINATION_QUERY } from '../components/Pagination'

/**
 * A helper function to create mocks for the given number of items
 * @param count The count of items to mock
 */
function paginationMocks(count) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: count,
              __typename: 'count',
            },
          },
        },
      },
    },
  ]
}

// Mock the NextJS router
Router.router = {
  push() {},
  prefetch() {},
}

describe('<Pagination />', () => {
  it('renders correct pagination for 10 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={paginationMocks(10)}>
        <Pagination page={1} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    const pagination = wrapper.find('div[data-test="pagination"]')
    expect(toJSON(pagination)).toMatchSnapshot()
    expect(pagination.text()).toContain('Page 1 of 3')
    expect(pagination.text()).toContain('10 Items Total')
  })
  it('disables prev button on first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={paginationMocks(10)}>
        <Pagination page={1} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    const prevLink = wrapper.find('a.prev')
    const nextLink = wrapper.find('a.next')
    expect(prevLink.prop('aria-disabled')).toBe(true)
    expect(nextLink.prop('aria-disabled')).toBe(false)
  })
  it('disables next button on last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={paginationMocks(10)}>
        <Pagination page={3} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    const prevLink = wrapper.find('a.prev')
    const nextLink = wrapper.find('a.next')
    expect(prevLink.prop('aria-disabled')).toBe(false)
    expect(nextLink.prop('aria-disabled')).toBe(true)
  })
  it('enables all buttons on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={paginationMocks(10)}>
        <Pagination page={2} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    const prevLink = wrapper.find('a.prev')
    const nextLink = wrapper.find('a.next')
    expect(prevLink.prop('aria-disabled')).toBe(false)
    expect(nextLink.prop('aria-disabled')).toBe(false)
  })
})
