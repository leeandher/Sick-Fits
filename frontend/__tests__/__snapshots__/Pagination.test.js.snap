// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders correct pagination for 10 items 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <Pagination__StylishAnchor
      aria-disabled={true}
      className="prev"
      onClick={[Function]}
    >
      <a
        aria-disabled={true}
        className="prev Pagination__StylishAnchor-w8etnd-0 cWUbgn"
        onClick={[Function]}
      >
        ← Prev
      </a>
    </Pagination__StylishAnchor>
  </Link>
  <p>
    Page 
    1
     of 
    3
  </p>
  <p>
    10
     Items Total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <Pagination__StylishAnchor
      aria-disabled={false}
      className="next"
      onClick={[Function]}
    >
      <a
        aria-disabled={false}
        className="next Pagination__StylishAnchor-w8etnd-0 cWUbgn"
        onClick={[Function]}
      >
        Next →
      </a>
    </Pagination__StylishAnchor>
  </Link>
</div>
`;
